publisher/publisher.go

```diff
@@ -50,36 +47,44 @@ func init() {
 
 type Publisher interface {
 	// Publish accepts an io.Reader and scans it line-by-line, parses the
-	// relevant event from each line, and sends to the target (Honeycomb)
-	Publish(r io.Reader) error
+	// relevant event from each line (using EventParser), and sends to the
+	// target (Honeycomb).
+	Publish(f state.DownloadedObject) error
+}
+
+type EventParser interface {
+	// ParseEvents runs in a background goroutine and parses the downloaded
+	// object, sending the events parsed from it further down the pipeline
+	// using the output channel. er
+	ParseEvents(obj state.DownloadedObject, out chan<- event.Event) error
+
+	// DynSample dynamically samples events, reading them from `eventsCh`
+	// and sending them to `sampledCh`. Behavior is dependent on the
+	// publisher implementation, e.g., some fields might matter more for
+	// ELB than for CloudFront.
+	DynSample(in <-chan event.Event, out chan<- event.Event)
 }
 
 // HoneycombPublisher implements Publisher and sends the entries provided to
 // Honeycomb. Publisher allows us to have only one point of entry to sending
 // events to Honeycomb (if desired), as well as isolate line parsing, sampling,
 // and URL sub-parsing logic.
 type HoneycombPublisher struct {
-	APIHost      string
-	SampleRate   int
-	nginxParser  *nginx.Parser
-	lines        chan string
-	eventsToSend chan event.Event
-	sampler      dynsampler.Sampler
+	Stater              state.Stater
+	EventParser         EventParser
```

Comment:
there are embedded structs in several places in this repo - it's now weird that these two are called out with names while the others aren't. I rescind my request that they be named; embed away.

Reply:
cool, will do